src/Controllers/AdminToolbarActionController.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace WeDevelop\AdminToolbar\Controllers;
+
+use SilverStripe\CMS\Model\SiteTree;
+use SilverStripe\Control\Controller;
+use SilverStripe\Control\HTTPRequest;
+use SilverStripe\Control\HTTPResponse;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Security\SecurityToken;
+use SilverStripe\Versioned\Versioned;
+use WeDevelop\AdminToolbar\Menus\Page\MenuItems\ArchiveMenuItem;
+use WeDevelop\AdminToolbar\Menus\Page\MenuItems\UnpublishAndArchiveMenuItem;
+use WeDevelop\AdminToolbar\Menus\Page\MenuItems\UnpublishMenuItem;
+
+class AdminToolbarActionController extends Controller
+{
+    /** @config */
+    private static string $url_segment = 'admintoolbaraction';
+
+    /**
+     * @config
+     * @var array<string>
+     */
+    private static array $allowed_actions = [
+        'pageAction',
+    ];
+
+    private static array $unpublishActions = [
+        UnpublishMenuItem::ACTION,
+        UnpublishAndArchiveMenuItem::ACTION,
+    ];
+
+    private static array $archiveActions = [
+        UnpublishAndArchiveMenuItem::ACTION,
+        ArchiveMenuItem::ACTION,
+    ];
+
+    private static $successMessages = [
+        UnpublishMenuItem::ACTION => 'Page succesfully unpublished',
+        ArchiveMenuItem::ACTION => 'Page succesfully archived',
+        UnpublishAndArchiveMenuItem::ACTION => 'Page succesfully unpublished and archived',
+    ];
+
+    public function pageAction(HTTPRequest $request): HTTPResponse
+    {
+        $csrfToken = $request->getHeader('X-CSRF-Token');
+        if (!SecurityToken::inst()->check($csrfToken)) {
+            return $this->httpError(400, 'CSRF token mismatch');
+        }
```

Comment:
You can use the `SecurityToken` to direct check the request rather then unpack the token yourself.

```php
if (!SecurityToken::inst()->checkRequest($request)) {
    return $this->httpError(400, 'CSRF token mismatch');
}
```